### PR TITLE
ENG-9489 sql functions. DO NOT MERGE! FOR REVIEW ONLY!

### PR DESCRIPTION
**Pull request by paulmartel**
Apply contributed patch to 1a0c54a master.

### CODE REVIEW

File: build.py [hunk 1]
```diff
@@ -74,7 +74,7 @@
     CTX.CPPFLAGS += " -fvisibility=default -DPROFILE_ENABLED"
 
 # linker flags
-CTX.LDFLAGS += """ -g3"""
```
Comment (paulmartel) [inline build.py]: This linker option and the boost/regex.hpp include in stringfunctions.h need to be supported in a general way that works for both VoltDB's cross-platform build environment AND for VoltDB's community and commercial deployments.

The VoltDB code base to date has avoided dependency on boost library code by only using boost "header-only" features.

Any boost version dependencies need to be identified for the purpose of our internal cross-platform build and test procedures. Our build and test system tend to work in two modes:
  1. build and test on each supported platform. (including Centos and RedHat 5 and later and Ubuntu 12.04 and later, and OSX 10.6 and later)
  2. build on Centos 5 and test on compatible linux platforms (including Centos and RedHat 5 and later, and Ubuntu 10.04 and later)

Centos/RedHat 6 and Ubuntu 12.04 may become prerequisites for building or deploying VoltDB release 6 or greater.

Normally, for safety in product deployment, external open source library dependencies are resolved by version-controlled components included in the VoltDB source and binary packaging.

File: src/ee/expressions/datefunctions.h [hunk 2]
```diff
@@ -393,4 +402,280 @@ template<> inline NValue NValue::callConstant<FUNC_CURRENT_TIMESTAMP>() {
...
+    if (number.isNull()) {
+        return getNullValue();
+    }
+
+    const NValue& date = arguments[1];
+    if (date.isNull()) {
+        return getNullValue();
+    }
+
+    int64_t interval = number.castAsBigIntAndGetValue();
```
Comment (paulmartel) [inline src/ee/expressions/datefunctions.h]: The remainder of this function -- except for multiplying interval by 3 -- should be factored into a static function that is also called by <FUNC_VOLT_DATEADD_MONTH> for maximized code reuse.

File: src/ee/expressions/datefunctions.h [hunk 3]
```diff
@@ -393,4 +402,280 @@ template<> inline NValue NValue::callConstant<FUNC_CURRENT_TIMESTAMP>() {
...
+    if (interval > PTIME_MAX_YEARS || interval < PTIME_MIN_YEARS) {
+        throw SQLException(SQLException::data_exception_numeric_value_out_of_range, "interval is too large for DATEADD function");
+    }
+
+    int64_t epoch_micros = date.getTimestamp();
+    boost::gregorian::date as_date;
+    boost::posix_time::time_duration as_time;
+    micros_to_date_and_time(epoch_micros, as_date, as_time);
+    boost::posix_time::ptime ts(as_date, as_time);
```
Comment (paulmartel) [inline src/ee/expressions/datefunctions.h]: These five lines are often repeated -- they could be factored out into a module-static inline function.

File: src/ee/expressions/functionexpression.cpp [hunk 4]
```diff
@@ -202,6 +202,9 @@ ExpressionUtil::functionFactory(int functionId, const std::vector<AbstractExpres
         case FUNC_CURRENT_TIMESTAMP:
             ret = new ConstantFunctionExpression<FUNC_CURRENT_TIMESTAMP>();
             break;
+        case FUNC_PI:
+            ret = new ConstantFunctionExpression<FUNC_PI>();
```
Comment (paulmartel) [inline src/ee/expressions/functionexpression.cpp]: call<FUNC_PI> could be avoided entirely by substituting a constant value for pi in the expression tree.

    ret = new ConstantValueExpression(getDoubleValue(boost::math::constants::pi<double>()));

File: src/ee/expressions/functionexpression.cpp [hunk 5]
```diff
@@ -424,6 +427,36 @@ ExpressionUtil::functionFactory(int functionId, const std::vector<AbstractExpres
         case FUNC_VOLT_SUBSTRING_CHAR_FROM:
             ret = new GeneralFunctionExpression<FUNC_VOLT_SUBSTRING_CHAR_FROM>(*arguments);
             break;
+        case FUNC_VOLT_DATEADD_YEAR:
```
Comment (paulmartel) [inline src/ee/expressions/functionexpression.cpp]: This list of cases is intended to be organized alphabetically to avoid code merge conflicts.

File: src/ee/expressions/stringfunctions.h [hunk 6]
```diff
@@ -636,6 +637,81 @@ template<> inline NValue NValue::call<FUNC_VOLT_FORMAT_CURRENCY>(const std::vect
...
+    if (pat.isNull()) {
+        return getNullValue();
+    }
+    if (pat.getValueType() != VALUE_TYPE_VARCHAR) {
+        throwCastSQLException(pat.getValueType(), VALUE_TYPE_VARCHAR);
+    }
+
+    int32_t lenSource = source.getObjectLength_withoutNull();
+    char* sourceChars = reinterpret_cast<char*>(source.getObjectValue_withoutNull());
+    std::string sourceStr(sourceChars, lenSource);
```
Comment (paulmartel) [inline src/ee/expressions/stringfunctions.h]: The std::string sourceStr requires a buffer copy and is not required to get the equivalent of sourceStr.begin(), sourceStr.end(), and sourceStr.c_str(). Observe how call<FUNC_VOLT_SUBSTRING_CHAR_FROM> uses valueChars and valueEnd as an alternative to constructing a std::string when the underlying API does not require one.

File: src/ee/expressions/stringfunctions.h [hunk 7]
```diff
@@ -636,6 +637,81 @@ template<> inline NValue NValue::call<FUNC_VOLT_FORMAT_CURRENCY>(const std::vect
...
+            position = NValue::getCharLength(sourceStr.c_str(), rit->position()) + 1;
+        } else {
+            position = 0;
+        }
+    } catch (boost::regex_error& e) {
+        // TODO Return more specific error messages?
+        throw SQLException(SQLException::data_exception_invalid_parameter, "illegal pattern string");
+    }
+
+    return getIntegerValue(static_cast<int32_t>(position));
```
Comment (paulmartel) [inline src/ee/expressions/stringfunctions.h]: A sql function that is configured with a BIGINT result value should use getBigIntValue rather than getIntegerValue for consistency. I'd need to consider the minor trade-offs of whether to standardize on a BIGINT or INTEGER return type. As long as this sql function is made self-consistent, I'd expect the trade-offs between BIGINT and INTEGER to be minor.

File: src/frontend/org/voltdb/jdbc/JDBC4DatabaseMetaData.java [hunk 8]
```diff
@@ -614,7 +614,7 @@ public int getMaxUserNameLength() throws SQLException
     public String getNumericFunctions() throws SQLException
     {
         checkClosed();
-        return "ABS,BITAND,BITNOT,BITOR,BIT_SHIFT_LEFT,BIT_SHIFT_RIGHT,BITXOR,CEILING,EXP,FLOOR,LN,LOG,POWER,SQRT";
+        return "ABS,BITAND,BITNOT,BITOR,BIT_SHIFT_LEFT,BIT_SHIFT_RIGHT,BITXOR,CEILING,EXP,FLOOR,LN,LOG,PI,POWER,SQRT";
```
Comment (paulmartel) [inline src/frontend/org/voltdb/jdbc/JDBC4DatabaseMetaData.java]: The other two new function names should be added to other function lists in this module.

File: src/hsqldb19b3/org/hsqldb_voltpatches/FunctionForVoltDB.java [hunk 9]
```diff
@@ -224,6 +236,17 @@ private FunctionId(String name, Type type, int id, int typeParameter, Type[] par
                     new Type[] { Type.SQL_BIGINT },
                     new short[] {  Tokens.OPENBRACKET, Tokens.QUESTION, Tokens.CLOSEBRACKET }),
 
+           new FunctionId("regexp_position", Type.SQL_BIGINT, FUNC_VOLT_REGEXP_POSITION, -1,
```
Comment (paulmartel) [inline src/hsqldb19b3/org/hsqldb_voltpatches/FunctionForVoltDB.java]: What is the history and rationale for this function name?
Has a function by this exact name been previously implemented by some other popular database vendor, or is it a new invention? I had trouble finding one online.
If so, what exact regexp format(s) do they support?

I believe that there is a similar function in the SQL 2011 standard named "position_regex"
(words transposed and no final 'p'). 
It has a slightly different (keyword based) syntax and more extensive options. It uses XQuery regexp format which may not be 100% compatible with the boost normal (Perl format) regexp that is used here -- maybe that in itself is a reason not to use that standard name.

File: src/ee/expressions/stringfunctions.h [hunk 10]
```diff
@@ -636,6 +637,81 @@ template<> inline NValue NValue::call<FUNC_VOLT_FORMAT_CURRENCY>(const std::vect
...
+
+    int32_t lenSource = source.getObjectLength_withoutNull();
+    char* sourceChars = reinterpret_cast<char*>(source.getObjectValue_withoutNull());
+    std::string sourceStr(sourceChars, lenSource);
+
+    int32_t lenPat = pat.getObjectLength_withoutNull();
+    char* patChars = reinterpret_cast<char*>(pat.getObjectValue_withoutNull());
+    std::string patStr(patChars, lenPat);
+
+    boost::regex_constants::syntax_option_type matchOpts = boost::regex_constants::normal;
```
Comment (paulmartel) [inline src/ee/expressions/stringfunctions.h]: What is the rationale for using Perl format regexp's -- are there better alternatives?

Release 5.8 does not support C++ 11 features, but a future release (possibly 6.0) MAY support some C++ 11 features. C++ 11 specifically has some native regexp support functions. Would it be possible for us to eliminate the boost dependency and migrate to a native implementation at that point?
Would there be any noticeable behavior differences -- such as the supported regexp syntax.
It might make sense to "future-proof" the implementation by using a boost configuration that we know can be consistently re-implemented with the planned C++ 11 native support.

**Comment by paulmartel** [comment]
I have reviewed this changeset as a delta from master for expedience in providing feedback.
But my understanding is that the intended short-term target for these changes is a patch release to v5.8 from which master has already diverged. 
To avoid delays in making these changes available in a v5.8 patch release, it would be more appropriate to submit code changes to the latest release_5.8.x branch (currently 4f437d0).

**Comment by paulmartel** [comment]
The VoltDB build system enforces a small number of our coding standards.
One of them is the use of spaces instead of tabs in source code.
Another is the absence of any kind of whitespace at the end of source code lines.
There is an ant rule (in the voltdb directory, run "ant licensecheck") that checks for violations of these standards and lists the offending source files. This rule will immediately halt our automated test suites when it detects violations.
There is also an automated source code cleanup script that will correctly solve these problems, leaving neatly formatted source code and allowing "ant licensecheck" to then succeed in most cases. In the voltdb directory, run "tools/licensescheck.py --fixws".
Other useful ant rules to run prior to submitting changesets are "ant eecheck" which builds and runs a quick set of C++ unit tests -- and "ant check" which runs a longer battery of regression tests including the junit tests.

File: tests/frontend/org/voltdb/regressionsuites/TestFunctionsForVoltDBSuite.java [hunk 11]
```diff
@@ -2068,4 +2068,271 @@ public void testBin() throws NoConnectionsException, IOException, ProcCallExcept
             validateTableColumnOfScalarVarchar(result, new String[]{binString});
         }
     }
+    
+    public void testDateadd() throws NoConnectionsException, IOException, ProcCallException {
+        System.out.println("STARTING test DATEADD function tests");
```
Comment (paulmartel) [inline tests/frontend/org/voltdb/regressionsuites/TestFunctionsForVoltDBSuite.java]: Needs test coverage for NULL passed to either or both of bigint and timestamp argument.

File: tests/frontend/org/voltdb/regressionsuites/TestFunctionsForVoltDBSuite.java [hunk 12]
```diff
@@ -2068,4 +2068,271 @@ public void testBin() throws NoConnectionsException, IOException, ProcCallExcept
...
+        try {
+            client.callProcedure("@AdHoc", "SELECT REGEXP_POSITION(DESC, '[a-z](\\d+)[A-Z]', 'k') FROM P1 WHERE REGEXP_POSITION(DESC, '[a-z](\\d+)[A-Z]', 'k') > 0");
+        } catch (ProcCallException e) {
+            assertEquals(ClientResponse.GRACEFUL_FAILURE, e.getClientResponse().getStatus());
+            assertTrue(e.getClientResponse().getStatusString().contains("illegal match flags"));
+        }        
+        
+        cr = client.callProcedure("@AdHoc", "DELETE FROM P1 WHERE ID = 200");
+        assertEquals(ClientResponse.SUCCESS, cr.getStatus());
+    }
```
Comment (paulmartel) [inline tests/frontend/org/voltdb/regressionsuites/TestFunctionsForVoltDBSuite.java]: needs test coverage for null arguments.
needs test coverage for non-ascii characters in DESC and/or in the pattern.
verify that positions are returned as character positions, not byte positions after multi-byte characters.

File: src/ee/expressions/stringfunctions.h [hunk 13]
```diff
@@ -636,6 +637,81 @@ template<> inline NValue NValue::call<FUNC_VOLT_FORMAT_CURRENCY>(const std::vect
...
+            for(std::string::iterator it = flagStr.begin(); it != flagStr.end(); ++it) {
+                switch (*it) {
+                    case 'c':
+                        break;
+                    case 'i':
+                        matchOpts |= boost::regex_constants::icase;
+                        break;
+                    default:
+                        throw SQLException(SQLException::data_exception_invalid_parameter, "illegal match flags");
+                }
```
Comment (paulmartel) [inline src/ee/expressions/stringfunctions.h]: It should probably be an error to specify contradictory flags like 'ci' or 'ic'.
Possibly, it should also be an error to repeat flags like 'cc' or 'iii'
Where do the specifications come from for the set of supported match flag characters and their meanings? Are they intended to be case sensitive?

**Comment by paulmartel** [comment]
I have finished my initial review. Please let me know via e-mail or by commenting here if you have further questions.

**Comment by paulmartel** [comment]
Correction: one last review comment. These functions need some basic cases covered by c++ unit tests. See tests/ee/common/nvalue_test.cpp where there are tests for other sql functions.
